tests/lean/run/subobject_from_source.lean

```diff
@@ -0,0 +1,8 @@
+structure foo :=
+(a b : ℕ := 2)
+
+structure bar :=
+(foo : foo)
+(c : ℕ)
+
+example (b : bar) : b.foo = {c := 2, ..b}.foo := rfl
```

Comment:
I tried this test on the `master` branch, and it works.
I don't know what this commit is fixing. To be honest, I have a hard time reviewing commits that modify `visit_structure_instance`. This method has more than 300 lines. If you have time, please break it in smaller methods and add comments.

Reply:
> I don't know what this commit is fixing.

Previously the RHS expanded to `(bar.mk (foo.mk b.a b.b) 2).foo`. Now it expands to `(bar.mk b.foo 2).foo`, which is definitionally equal to the LHS. I'll expand the comment with an example.

> To be honest, I have a hard time reviewing commits that modify visit_structure_instance.

Yeah, it's a mess. I dreaded splitting it into multiple functions because there are so many shared variables. Should I refactor it into a friend class of `elaborator` instead?

Reply:
> Previously the RHS expanded to (bar.mk (foo.mk b.a b.b) 2).foo. Now it expands to (bar.mk b.foo 2).foo, which is definitionally equal to the LHS. I'll expand the comment with an example.

I'm confused, in the current master (without this PR), I'm already getting `(bar.mk b.foo 2).foo`.
I just tried
```lean
def ex (b : bar) : b.foo = {c := 2, ..b}.foo := rfl

set_option pp.structure_instances false
#print ex
```
and I get
```lean
def ex : ∀ (b : bar), b.foo = (bar.mk (b.foo) 2).foo :=
λ (b : bar), rfl
```

Reply:
> Yeah, it's a mess. I dreaded splitting it into multiple functions because there are so many shared variables. Should I refactor it into a friend class of elaborator instead?

Yes, I think a fried class is the best option is this kind of situation.

Reply:
> in the current master (without this PR), I'm already getting (bar.mk b.foo 2).foo.

That's what I get for writing the test after the fix, thanks for catching that. `bar` should of course extend `foo`. Fixed.